hikari/impl/rest.py

```diff
@@ -425,6 +422,61 @@ def acquire(
         return rest_client
 
 
+@attr.define()
+class _LiveAttributes:
+    """Fields which are only present within `RESTClientImpl` while it's "alive".
+
+    !!! note
+        This must be started within an active asyncio event loop.
+    """
+
+    buckets: buckets_.RESTBucketManager
+    client_session: aiohttp.ClientSession
+    closed_event: asyncio.Event
+    # We've been told in DAPI that this is per token.
+    global_rate_limit: rate_limits.ManualRateLimiter
+    tcp_connector: aiohttp.TCPConnector
+
+    @classmethod
+    def build(
+        cls, max_rate_limit: float, http_settings: config.HTTPSettings, proxy_settings: config.ProxySettings
+    ) -> _LiveAttributes:
+        """Build a live attributes object.
+
+        !!! warning
+            This can only be called when the current thread has an active
+            asyncio loop.
+        """
+        # This asserts that this is called within an active event loop.
+        asyncio.get_running_loop()
+        tcp_connector = net.create_tcp_connector(http_settings)
+        _LOGGER.log(ux.TRACE, "acquired new tcp connector")
+        client_session = net.create_client_session(
+            connector=tcp_connector,
+            # No, this is correct. We manage closing the connector ourselves in this class.
+            # This works around some other lifespan issues.
+            connector_owner=False,
+            http_settings=http_settings,
+            raise_for_status=False,
+            trust_env=proxy_settings.trust_env,
+        )
+        _LOGGER.log(ux.TRACE, "acquired new aiohttp client session")
+        return _LiveAttributes(
+            buckets=buckets_.RESTBucketManager(max_rate_limit),
+            client_session=client_session,
+            closed_event=asyncio.Event(),
+            global_rate_limit=rate_limits.ManualRateLimiter(),
+            tcp_connector=tcp_connector,
+        )
+
+    async def close(self) -> None:
+        self.closed_event.set()
+        self.buckets.close()
+        await self.client_session.close()
+        self.global_rate_limit.close()
+        await self.tcp_connector.close()
```

Comment:
Cant we have all this logic inside the `RESTClientImpl`?

Reply:
Sure but I prefer the separation of having the logic which manages those fields be in the same place they're stored, kinda helps with readability imo tbh

Reply:
Having all the `_get_live_attributes()` calls doesnt look nice to read (imo)

Reply:
Less calls can be made to that if LiveAttributes itself internally tracks whether it's been closed or not

Reply:
Fwiw I've pushed another approach which lets LiveAttributes track whether it's been told to close or not internally, this comes with the added benefit of letting us distinguish between when a call fails because the client was closed and when a call fails because the client was closed as the request was being made in the error message